ganga/GangaCore/GPIDev/Base/Proxy.py

```diff
@@ -1205,6 +1205,165 @@ def _getattribute(self, name):
+                return
+
+            setattr(object.__getattribute__(self, "_obj"), name, value)
+            if name in attr_map:
+                attr_map[name] = value
+            object.__setattr__(self, "_attr_map", attr_map)
+        else:
+            object.__setattr__(self, name, value)
+
+    _special_names = [
+        '__abs__', '__add__', '__and__', '__call__', '__cmp__', '__coerce__',
+        '__contains__', '__delitem__', '__delslice__', '__div__', '__divmod__',
+        '__eq__', '__float__', '__floordiv__', '__ge__', '__getitem__',
+        '__getslice__', '__gt__', '__hash__', '__hex__', '__iadd__', '__iand__',
+        '__idiv__', '__idivmod__', '__ifloordiv__', '__ilshift__', '__imod__',
+        '__imul__', '__int__', '__invert__', '__ior__', '__ipow__', '__irshift__',
+        '__isub__', '__iter__', '__itruediv__', '__ixor__', '__le__', '__len__',
+        '__long__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__',
+        '__neg__', '__nonzero__', '__oct__', '__or__', '__pos__', '__pow__',
+        '__radd__', '__rand__', '__rdiv__', '__rdivmod__', '__reduce__', '__reduce_ex__',
+        '__repr__', '__reversed__', '__rfloorfiv__', '__rlshift__', '__rmod__',
+        '__rmul__', '__ror__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__',
+        '__rtruediv__', '__rxor__', '__setitem__', '__setslice__', '__str__',
+        '__sub__', '__truediv__', '__xor__', 'next',
+    ]
+
+    @classmethod
+    def _create_class_proxy(cls, theclass):
+        """creates a proxy for the given class"""
+
+        def make_method(name):
+            def method(self, *args, **kw):
+                if name in cls._special_names and args is not ():
+                    args = map(lambda x: x._obj if isinstance(x, COWProxy) else x, args)
+                return getattr(object.__getattribute__(self, "_obj"), name)(*args, **kw)
+            return method
+
+        namespace = {}
+        for name in cls._special_names:
+            if hasattr(theclass, name):
+                namespace[name] = make_method(name)
+        return type("%s(%s)" % (cls.__name__, theclass.__name__), (cls,), namespace)
+
+    def __new__(cls, obj, *args, **kwargs):
+        """
+        creates an proxy instance referencing `obj`. (obj, *args, **kwargs) are
+        passed to this class' __init__, so deriving classes can define an
+        __init__ method of their own.
+        note: _class_proxy_cache is unique per deriving class (each deriving
+        class must hold its own cache)
+        """
+        try:
+            cache = cls.__dict__["_class_proxy_cache"]
+        except KeyError:
+            cls._class_proxy_cache = cache = {}
+        try:
+            theclass = cache[obj.__class__]
+        except KeyError:
+            cache[obj.__class__] = theclass = cls._create_class_proxy(obj.__class__)
+        ins = object.__new__(theclass)
+        theclass.__init__(ins, obj, *args, **kwargs)
+        return ins
+
+
```

Comment:
My general feeling is that this is quite dense, especially with all the static calls to the object base class. I through together a simple first approach. What are the deficiencies of the following?
```python
class COWProxy(object):
    __slots__ = ('_obj', '_partial', '_partial_dict', '_copied')

    def __init__(self, obj, partial=False):
        self._obj = obj
        self._partial = partial
        self._partial_dict = {}
        self._copied = False

    def __getattribute__(self, attr):
        if self._copied:
            return getattr(self._obj, attr)
        return self._partial_dict.setdefault(attr, COWProxy(getattr(self._obj, attr)))

    def __setattr__(self, attr, value):
        if self._copied:
            return setattr(self._obj, attr, value)
        if not self._partial:
            self._obj = deepcopy(self._obj)
            self._copied = True
            return setattr(self._obj, attr, value)
        self._partial_dict[attr] = value
```